File: app/src/main/java/com/door43/translationstudio/core/MergeConflictsHandler.java
package com.door43.translationstudio.core;

import android.os.Handler;
import android.os.Looper;

import com.door43.translationstudio.App;
import com.door43.usecases.ParseMergeConflicts;

import org.unfoldingword.tools.taskmanager.ManagedTask;
import org.unfoldingword.tools.taskmanager.TaskManager;

import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Created by blm on 11/22/16.
 */

public class MergeConflictsHandler {
    public static final String mergeConflictHead = "<<<<<<< HEAD.*\\n";
    public static Pattern mergeConflictPatternHead = Pattern.compile(mergeConflictHead);


    /**
     * Split the merge conflict into a list of the options
     *
     * @param text
     * @return
     */
    static public List<CharSequence> getMergeConflictItems(String text) {
        return ParseMergeConflicts.INSTANCE.execute(text);
    }

    /**
     * Split the merge conflict into a list of the options
     *
     * @param text
     * @return
     */
    static public CharSequence getMergeConflictItemsHead(String text) {
        List<CharSequence> items = getMergeConflictItems(text);
        if(items.isEmpty()) {
            return null;
        }
        return items.get(0);
    }

    /**
     * Detects merge conflict tags
     *
     * @param text
     * @return
     */
    static public boolean isMergeConflicted(CharSequence text) {
        if ((text != null) && (text.length() > 0)) {
            Matcher matcher = mergeConflictPatternHead.matcher(text);
            return matcher.find();
        }
        return false;
    }

    /**
     * search for first merge conflict - We need this to double check that there is a conflict in any chunks
     *
     * @param targetTranslationId
     * @return
     */
    static public boolean isTranslationMergeConflicted(String targetTranslationId) {
        if(targetTranslationId == null) {
            return false;
        }

        Translator translator = App.getTranslator();

        TargetTranslation targetTranslation = translator.getTargetTranslation(targetTranslationId);
        if(targetTranslation == null) {
            return false;
        }

        ProjectTranslation pt = targetTranslation.getProjectTranslation();
        if(pt == null) {
            return false;
        }

        if(isMergeConflicted(pt.getTitle())) {
            return true;
        }

        ChapterTranslation[] chapters = targetTranslation.getChapterTranslations();
        for(ChapterTranslation ct:chapters) {
            if(isMergeConflicted(ct.title)) {
                return true;
            }

            if(isMergeConflicted(ct.reference)) {
                return true;
            }

            FrameTranslation[] frames = targetTranslation.getFrameTranslations(ct.getId(), TranslationFormat.DEFAULT);
            for (FrameTranslation frame : frames) {
                if (isMergeConflicted(frame.body)) {
                    return true;
                }
            }
        }

        return false;
    }

    /**
     * check the whole project to see if there is actually a chunk conflict
     * @param targetTranslationId
     * @param listener
     */
    public static void backgroundTestForConflictedChunks(final String targetTranslationId, final OnMergeConflictListener listener) {
        ManagedTask task = new ManagedTask() {
            @Override
            public void start() {
                try {
                    if(interrupted()) return;
                    boolean conflicted = MergeConflictsHandler.isTranslationMergeConflicted(targetTranslationId);
                    setResult(conflicted);
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        };
        task.addOnFinishedListener(task1 -> {
            TaskManager.clearTask(task1);
            boolean conflicted = false;
            if(task1.getResult() != null) conflicted = (boolean) task1.getResult();
            if(!task1.isCanceled()) {
                Handler hand = new Handler(Looper.getMainLooper());
                final boolean finalConflicted = conflicted;
                hand.post(() -> {
                    if(finalConflicted) {
                        listener.onMergeConflict(targetTranslationId);
                    } else {
                        listener.onNoMergeConflict(targetTranslationId);
                    }
                });
            }
        });
        TaskManager.addTask(task);
    }

    public interface OnMergeConflictListener {
        void onNoMergeConflict(String targetTranslationId);
        void onMergeConflict(String targetTranslationId);
    }
}

File: app/src/main/java/com/door43/translationstudio/core/MergeConflictsHandler.kt
package com.door43.translationstudio.core

import com.door43.usecases.ParseMergeConflicts

/**
 * Created by blm on 11/22/16.
 */
object MergeConflictsHandler {
    private const val MERGE_CONFLICT_HEAD = "<<<<<<< HEAD.*\\n"
    private val mergeConflictPatternHead: Regex = MERGE_CONFLICT_HEAD.toRegex()

    /**
     * Split the merge conflict into a list of the options
     *
     * @param text
     * @return
     */
    fun getMergeConflictItems(text: String): List<CharSequence> {
        // Assuming ParseMergeConflicts is a Kotlin object or has a static execute method
        return ParseMergeConflicts.execute(text)
    }

    /**
     * Split the merge conflict into a list of the options
     *
     * @param text
     * @return
     */
    fun getMergeConflictItemsHead(text: String): CharSequence? {
        val items = getMergeConflictItems(text)
        return if (items.isEmpty()) null else items[0]
    }

    /**
     * Detects merge conflict tags
     *
     * @param text
     * @return
     */
    fun isMergeConflicted(text: CharSequence): Boolean {
        if (text.isNotEmpty()) {
            return mergeConflictPatternHead.containsMatchIn(text)
        }
        return false
    }

    /**
     * search for first merge conflict - We need this to double-check that there is a conflict in any chunks
     *
     * @param targetTranslationId
     * @return
     */
    fun isTranslationMergeConflicted(
        targetTranslationId: String?,
        translator: Translator
    ): Boolean {
        if (targetTranslationId == null) {
            return false
        }

        val targetTranslation = translator.getTargetTranslation(targetTranslationId) ?: return false

        val pt = targetTranslation.projectTranslation
        if (isMergeConflicted(pt.title)) {
            return true
        }

        val chapters = targetTranslation.chapterTranslations
        for (ct in chapters) {
            if (isMergeConflicted(ct.title)) {
                return true
            }

            if (isMergeConflicted(ct.reference)) {
                return true
            }

            val frames = targetTranslation.getFrameTranslations(
                ct.id,
                TranslationFormat.DEFAULT
            )
            for (frame in frames) {
                if (isMergeConflicted(frame.body)) {
                    return true
                }
            }
        }

        return false
    }
}